watcher.go

```diff
@@ -148,25 +149,52 @@ func (w *Watcher) waitForShutdown(stop <-chan struct{}) {
 	}()
 }
 
-func (w *Watcher) getOrCreatePromSDConfigMap(client *kubernetes.Clientset) (*v1.ConfigMap, error) {
-	configMap, err := client.CoreV1().ConfigMaps("istio-system").
+// discover Prometheus deployment namespace and create ConfigMap
+func (w *Watcher) getOrCreatePromSDConfigMap(client *kubernetes.Clientset) (*v1.ConfigMap, string, error) {
+	if w.namespace == "" {
+		var err error
+		w.namespace, err = discoverPromNamespace(client)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+
+	configMap, err := client.CoreV1().ConfigMaps(w.namespace).
 		Get(context.TODO(), "file-sd-config", metav1.GetOptions{})
 	if err == nil {
 		// config map exists, return directly
-		return configMap, nil
+		return configMap, w.namespace, nil
 	}
+	
 	cfg := &v1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: "file-sd-config",
 		},
 		Data: make(map[string]string),
 	}
 	cfg.Data[w.sdFileName] = ""
-	if configMap, err = client.CoreV1().ConfigMaps("istio-system").Create(context.TODO(), cfg,
+	if configMap, err = client.CoreV1().ConfigMaps(w.namespace).Create(context.TODO(), cfg,
 		metav1.CreateOptions{}); err != nil {
-		return nil, err
+		return nil, "", err
+	}
+	return configMap, w.namespace, nil
+}
+
+func discoverPromNamespace(client *kubernetes.Clientset) (string, error) {
+	label := "prometheus"
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"app": label}}
+	listOptions := metav1.ListOptions{
+		LabelSelector: labels.Set(labelSelector.MatchLabels).String(),
 	}
-	return configMap, nil
+
+	podsList, err := client.CoreV1().Pods("").List(context.TODO(), listOptions)
```

Comment:
this operation is both expensive (listing all pods in cluster) and not accurate- there could be many Prometheus running.

Should we just have a config option?

Reply:
I've tried list all namespace, but the client do not have permission to list all namespaces.  We only need to mount the fild-sd-config in one namespace and the watcher will keep watching all newly added vm.

The limitation of this implementation is the app name must be "prometheus".

For the config option, do you mean it's possible to include one more field to indicate prometheus namespace in the container config?

Reply:
@howardjohn i think taking a config option is the right ultimate approach, but for default behavior, trying to find running prometheuses isn't all that problematic, as a one-time operation. 

@xulingqing can we move the ns discovery to a one-time startup operation (outside of the watch), and support multiple namespaces ?

Reply:
> @xulingqing can we move the ns discovery to a one-time startup operation (outside of the watch), and support multiple namespaces ?

do you mean create config map in all namespaces that have prometheus deployment?

Reply:
Done move the ns discovery to a one-time startup operation(at the beginning of `start` watcher function).